### PR TITLE
feat!: Remove deprecated context methods

### LINKED ISSUE #17520: Change Request: Remove deprecated context pass-through SourceCode methods
### ESLint version

HEAD

### What problem do you want to solve?

We have a bunch of deprecated methods on `context` that pass through to `SourceCode`:

* `getComments()`
* * `getSource`
* `getAllComments`
* `getNodeByRangeIndex`
* `getComments`
* `getCommentsBefore`
* `getCommentsAfter`
* `getCommentsInside`
* `getJSDocComment`
* `getFirstToken`
* `getFirstTokens`
* `getLastToken`
* `getLastTokens`
* `getTokenAfter`
* `getTokenBefore`
* `getTokenByRangeStart`
* `getTokens`
* `getTokensAfter`
* `getTokensBefore`
* `getTokensBetween`

### What do you think is the correct solution?

I think it's time to remove these methods to clean up the codebase.

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Removed the deprecated methods on the `context` object.

fixes #17520

### CODE REVIEW

File: lib/linter/linter.js [hunk 1]
```diff
@@ -914,42 +914,7 @@ function createRuleListeners(rule, ruleContext) {
...
-        (contextInfo, methodName) =>
-            Object.assign(contextInfo, {
-                [methodName](...args) {
-                    return this.sourceCode[DEPRECATED_SOURCECODE_PASSTHROUGHS[methodName]](...args);
-                }
-            }),
-        {}
-    )
-);
+const BASE_TRAVERSAL_CONTEXT = {};
```
Comment (mdjermanovic) [inline lib/linter/linter.js:917]: Do we still need `BASE_TRAVERSAL_CONTEXT`? Looks like we can just remove it now.

**Comment by mdjermanovic** [comment]
We should also remove deprecation warnings from the RuleTester:

https://github.com/eslint/eslint/blob/5454c22b24f39be2dac7f28cfcfdb6c753faaf4e/lib/rule-tester/rule-tester.js#L175-L195

File: lib/linter/linter.js (hunk 1 above)
Comment (nzakas) [reply lib/linter/linter.js:917]: Ah, right you are!

**Comment by nzakas** [comment]
Oops, yes. Removed from the tests but not `RuleTester` itself. Good catch.

File: tests/lib/linter/linter.js [hunk 2]
```diff
@@ -10221,507 +9939,6 @@ describe("Linter with FlatConfigArray", () => {
...
-                        rules: { "test/checker": "error" }
-                    };
-
-                    linter.verify(code, config);
-                    assert(spy && spy.calledOnce);
-                });
-
-            });
-
-            describe("context.getAncestors()", () => {
```
Comment (mdjermanovic) [inline tests/lib/linter/linter.js:10520]: Tests for `context.getAncestors()` shouldn't be removed in this PR?

Comment (nzakas) [reply tests/lib/linter/linter.js:10520]: It looks like `getAncestors()`, `getScope()`, `getDeclaredVariables()`, and `markVariableAsUsed()` don't actually have a separate issue tracking their removal, despite us [announcing it](https://eslint.org/blog/2023/09/preparing-custom-rules-eslint-v9/).

So maybe I should remove those in this PR too?

Comment (mdjermanovic) [reply tests/lib/linter/linter.js:10520]: It's a subtask of https://github.com/eslint/eslint/issues/16999 (Phase 8: Removal of old APIs; just it seems that `markVariableAsUsed()` was accidentally omitted from that list, and note that the list includes removing `context.parserServices` as well) which is tracked in the v9 project.

Regardless, seems best and easiest for maintaining the alpha release to remove them as well in this PR 👍

Comment (nzakas) [reply tests/lib/linter/linter.js:10520]: Ah good call, those subtasks are tricky to remember. I've updated the list to reflect reality and will add the rest into this PR.

**Comment by nzakas** [comment]
Note: Linting is failed because `eslint-plugin-n` is still using the old `getScope()` and `getAncestors()`, which have been removed. I opened an issue here:
https://github.com/eslint-community/eslint-plugin-n/issues/143

**Comment by aladdin-add** [comment]
I've made a few updates:

* #17836 
* #17837 
* #17838

**Comment by mdjermanovic** [comment]
> I've made a few updates, hopefully it can fix the lint failing:
> 
> * [chore: update eslint-plugin-n v16.4.0 #17836](https://github.com/eslint/eslint/pull/17836)
> * [chore: upgrade eslint-plugin-unicorn to v49.0.0 #17837](https://github.com/eslint/eslint/pull/17837)
> * [chore: upgrade eslint-plugin-eslint-plugin v5.2.1 #17838](https://github.com/eslint/eslint/pull/17838)

All merged now. I checked this branch locally and linting works now. There are just several lint errors and a merge conflict in package.json that should be fixed.

**Comment by nzakas** [comment]
Fixed merge conflicts and re-pushed the branch.

**Comment by aladdin-add** [comment]
seems there are a few linting errors that need to be fixed.

**Comment by nzakas** [comment]
Thanks. Fixed!

File: docs/src/extend/custom-rules.md [hunk 3]
```diff
@@ -551,6 +537,20 @@ Once you have an instance of `SourceCode`, you can use the following methods on
...
+    * If the node is an `ArrowFunctionExpression`, variables for the parameters are returned.
+    * If the node is a `ClassDeclaration` or a `ClassExpression`, the variable for the class name is returned.
+    * If the node is a `CatchClause`, the variable for the exception is returned.
+    * If the node is an `ImportDeclaration`, variables for all of its specifiers are returned.
+    * If the node is an `ImportSpecifier`, `ImportDefaultSpecifier`, or `ImportNamespaceSpecifier`, the declared variable is returned.
+    * Otherwise, if the node does not declare any variables, an empty array is returned.
+* `getScope()`: Returns the [scope](./scope-manager-interface#scope-interface) of the currently-traversed node. This information can be used to track references to variables.
+* `markVariableAsUsed(name)`: Marks a variable with the given name in the current scope as used. This affects the [no-unused-vars](../rules/no-unused-vars) rule. Returns `true` if a variable with the given name was found and marked as used, otherwise `false`.
+
```
Comment (aladdin-add) [inline docs/src/extend/custom-rules.md:553]:
```suggestion
```

File: docs/src/extend/custom-rules.md [hunk 4]
```diff
@@ -551,6 +537,19 @@ Once you have an instance of `SourceCode`, you can use the following methods on
...
+    * If the node is a `VariableDeclaration`, all variables declared in the declaration are returned.
+    * If the node is a `VariableDeclarator`, all variables declared in the declarator are returned.
+    * If the node is a `FunctionDeclaration` or `FunctionExpression`, the variable for the function name is returned, in addition to variables for the function parameters.
+    * If the node is an `ArrowFunctionExpression`, variables for the parameters are returned.
+    * If the node is a `ClassDeclaration` or a `ClassExpression`, the variable for the class name is returned.
+    * If the node is a `CatchClause`, the variable for the exception is returned.
+    * If the node is an `ImportDeclaration`, variables for all of its specifiers are returned.
+    * If the node is an `ImportSpecifier`, `ImportDefaultSpecifier`, or `ImportNamespaceSpecifier`, the declared variable is returned.
+    * Otherwise, if the node does not declare any variables, an empty array is returned.
+* `getScope()`: Returns the [scope](./scope-manager-interface#scope-interface) of the currently-traversed node. This information can be used to track references to variables.
```
Comment (mdjermanovic) [inline docs/src/extend/custom-rules.md:551]:
```suggestion
* `getScope(node)`: Returns the [scope](./scope-manager-interface#scope-interface) of the given node. This information can be used to track references to variables.
```

Also, we can remove this line:

https://github.com/eslint/eslint/blob/ba87a0651a65b52c3ac442b512dd9f4c2b4c5f57/docs/src/extend/custom-rules.md?plain=1#L790

File: docs/src/extend/custom-rules.md [hunk 5]
```diff
@@ -551,6 +537,19 @@ Once you have an instance of `SourceCode`, you can use the following methods on
...
+* `markVariableAsUsed(name)`: Marks a variable with the given name in the current scope as used. This affects the [no-unused-vars](../rules/no-unused-vars) rule. Returns `true` if a variable with the given name was found and marked as used, otherwise `false`.
```
Comment (mdjermanovic) [inline docs/src/extend/custom-rules.md:552]:
```suggestion
* `markVariableAsUsed(name, refNode)`: Marks a variable with the given name in a scope indicated by the given reference node as used. This affects the [no-unused-vars](../rules/no-unused-vars) rule. Returns `true` if a variable with the given name was found and marked as used, otherwise `false`.
```

Also, we can remove this line:

https://github.com/eslint/eslint/blob/ba87a0651a65b52c3ac442b512dd9f4c2b4c5f57/docs/src/extend/custom-rules.md?plain=1#L839

File: tests/lib/linter/linter.js [hunk 6]
```diff
@@ -349,4451 +284,3751 @@ describe("Linter", () => {
...
-            linter.defineParser("enhanced-parser", testParsers.enhancedParser);
             const messages = linter.verify("0", config, filename);
             const suppressedMessages = linter.getSuppressedMessages();
 
-            assert.strictEqual(messages.length, 0);
+            assert.strictEqual(messages[0].message, filename);
             assert.strictEqual(suppressedMessages.length, 0);
         });
 
-        it("should expose parser services when using parseForESLint() and services are specified", () => {
```
Comment (mdjermanovic) [inline tests/lib/linter/linter.js:1195]: I think we should keep tests with `parserServices` as they're testing how Linter creates SourceCode objects based on parser features. We could just remove the `assert.strictEqual(context.parserServices, context.sourceCode.parserServices)` line.

File: tests/lib/linter/linter.js [hunk 7]
```diff
@@ -349,4451 +284,3751 @@ describe("Linter", () => {
...
             const messages = linter.verify("0", config, filename);
             const suppressedMessages = linter.getSuppressedMessages();
 
-            assert.strictEqual(messages.length, 1);
-            assert.strictEqual(messages[0].message, "Hi!");
+            assert.strictEqual(messages[0].message, filename);
             assert.strictEqual(suppressedMessages.length, 0);
         });
 
-        it("should use the same parserServices if source code object is reused", () => {
```
Comment (mdjermanovic) [inline tests/lib/linter/linter.js:1218]: Same as the previous comment, I think we should keep this test here.

File: tests/lib/linter/linter.js [hunk 8]
```diff
@@ -5297,6768 +4532,4466 @@ var a = "test2";
...
 
-                    assert.strictEqual(messages.length, 0);
-                    assert.strictEqual(suppressedMessages.length, 0);
-                });
+                assert.strictEqual(messages.length, 1);
+                assert.strictEqual(messages[0].severity, 2);
+                assert.isTrue(messages[0].fatal);
+                assert.match(messages[0].message, /^Parsing error:.*'enum'/u);
 
-                it("should expose parser services when using parseForESLint() and services are specified", () => {
```
Comment (mdjermanovic) [inline tests/lib/linter/linter.js:8909]: Same as the previous comment, I think we should keep this test here.

File: tests/lib/linter/linter.js [hunk 9]
```diff
@@ -5297,6768 +4532,4466 @@ var a = "test2";
...
+                assert.strictEqual(suppressedMessages.length, 0);
+            });
+        });
 
-                    assert.strictEqual(suppressedMessages.length, 0);
-                });
+        it("should not allow `allowReserved: true` in ES > 3", () => {
+            const ecmaVersions = [void 0, ...espree.supportedEcmaVersions.filter(ecmaVersion => ecmaVersion > 3)];
 
-                it("should use the same parserServices if source code object is reused", () => {
```
Comment (mdjermanovic) [inline tests/lib/linter/linter.js:8946]: Same as the previous comment, I think we should keep this test here.

File: tests/lib/linter/linter.js
Comment (mdjermanovic) [inline tests/lib/linter/linter.js:1]: Just a tip if anyone else is going to review this: better use VS Code diff viewer for this file, as this diff here is pretty much unusable and the diff is not as large as it appears to be.

**Comment by nzakas** [comment]
Also updated `eslint-plugin-n` so linting will no longer fail. 🎉

File: tests/lib/linter/linter.js [hunk 10]
```diff
@@ -349,3875 +284,3623 @@ describe("Linter", () => {
...
+            const config = { rules: { "test-service-rule": 2 }, parser: "enhanced-parser" };
+            const messages = linter.verify("0", config, filename);
+            const suppressedMessages = linter.getSuppressedMessages();
 
+            assert.strictEqual(messages.length, 1);
+            assert.strictEqual(messages[0].message, "Hi!");
+            assert.strictEqual(suppressedMessages.length, 0);
+        });
 
-    describe("when calling context.getScope", () => {
```
Comment (mdjermanovic) [inline tests/lib/linter/linter.js:613]: I just noticed that this set of tests wasn't copied into source-code tests, but we could do that in another PR.

File: tests/lib/linter/linter.js [hunk 11]
```diff
@@ -5490,6575 +4776,4351 @@ var a = "test2";
...
-            assert.strictEqual(scope.upper.type, "global");
-            assert.strictEqual(scope.block, node);
-            assert.strictEqual(scope.upper.variables[0].references.length, 0);
-            assert.strictEqual(scope.references[0].identifier, node.left.declarations[0].id);
-            assert.strictEqual(scope.references[1].identifier, node.right);
-            assert.strictEqual(scope.references[1].resolved, scope.variables[0]);
-        });
-    });
-
-    describe("Variables and references", () => {
```
Comment (mdjermanovic) [inline tests/lib/linter/linter.js:7218]: This set of tests also doesn't exist in source-code tests, though these look more like tests for eslint-scope than tests for linter.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!
